index.js

```diff
@@ -79,67 +87,133 @@ module.exports = {
     }
 
     if (type === 'app-boot') {
-      return fastbootAppModule(config.modulePrefix);
+      return fastbootAppModule(config.modulePrefix, JSON.stringify(config.APP || {}));
     }
 
-    if (type === 'config-module' && this.app.options.__is_building_fastboot__) {
-      var linesToRemove = contents.length;
-      while(linesToRemove) {
-        // Clear out the default config from ember-cli
-        contents.pop();
-        linesToRemove--;
-      }
-
-      return 'return FastBoot.config();';
+    // if the fastboot addon is installed, we overwrite the config-module so that the config can be read
+    // from meta tag for browser build and from Fastboot config for fastboot target
+    if (type === 'config-module') {
+      var emberCliPath = path.join(this.app.project.nodeModulesPath, 'ember-cli');
+      contents.splice(0, contents.length);
+      contents.push('if (typeof FastBoot !== \'undefined\') {');
+      contents.push('return FastBoot.config();');
+      contents.push('} else {');
+      contents.push('var prefix = \'' + config.modulePrefix + '\';');
+      contents.push(fs.readFileSync(path.join(emberCliPath, 'lib/broccoli/app-config-from-meta.js')));
+      contents.push('}');
+      return;
     }
   },
 
-  treeForApp: function(defaultTree) {
-    var trees = [defaultTree];
-
+  addToFastBootTree: function() {
+    // check the ember-cli version and conditionally patch the DOM api
     if (this._getEmberVersion().lt('2.10.0-alpha.1')) {
-      trees.push(this.treeGenerator(path.resolve(this.root, 'app-lt-2-9')));
+      return fastbootBuilder.addFastBootPath(path.resolve(__dirname, 'fastboot-app-lt-2-9'));
     }
-
-    return mergeTrees(trees, { overwrite: true });
   },
 
   /**
-   * Filters out initializers and instance initializers that should only run in
-   * browser mode.
+   * Function that builds the fastboot tree from all fastboot complaint addons
+   * and project and transpiles it into appname-fastboot.js
    */
-  preconcatTree: function(tree) {
-    return filterInitializers(tree, this.app.name);
+  _getFastbootTree: function() {
+    var appName = this._name;
+    var nodeModulesPath = this.project.nodeModulesPath;
+
+    // set the app name before building
+    fastbootBuilder.setAppName(appName);
+
+    var fastbootTrees = [];
+    this.project.addons.forEach((addon) => {
+      // walk through each addon and grab its fastboot tree
+      var currentAddonFastbootPath = path.join(nodeModulesPath, addon.name, 'fastboot');
+      if (existsSync(currentAddonFastbootPath)) {
+        var fastbootTree = fastbootBuilder.addFastBootPath(currentAddonFastbootPath);
+        fastbootTrees.push(fastbootTree);
+      }
+
+      // invoke addToFastBootTree for every addon
+      if (addon.addToFastBootTree) {
+        var additionalFastBootTree = addon.addToFastBootTree();
```

Comment:
@stefanpenner I actually like the idea of having `treeForFastBoot` which is more aligned to other existing APIs. And yes it would cover the usecase I wanted it to cover as well. Will update it to providing `treeForFastBoot` :)